contents/codeQuality/CodeQualityMetrics.md

```diff
@@ -0,0 +1,98 @@
+- It is, however, disputable how "simple", "understandable", etc. is defined, because they are all qualitative. What is "..." for one person may not be so "..." for another; in fact, it may not be so for that person three months later.
+
+A quantitative approach needs to be adopted to measure those qualities precisely. **Code Quality Metrics** is the answer. It provides a **quantitative measure** of the quality of a piece of code.
+
+> Since quantitative measurements are essential in all sciences, there is a continuous effort by computer science practitioners and theoreticians to bring similar approaches to software development.
+> The goal is obtaining objective, reproducible and quantifiable measurements, which may have numerous valuable applications in schedule and budget planning, cost estimation, quality assurance testing, software debugging, software performance optimization, and optimal personnel task assignments.
+>
+> https://en.wikibooks.org/wiki/Introduction_to_Software_Engineering/Quality/Metrics
+
+In fact, [the Boeing 777 project attributed its success to excellent usage of Quality Metrics (both code and non-code)](https://thesai.org/Downloads/Volume3No1/Paper%2021-Survey%20on%20Impact%20of%20Software%20Metrics%20on%20Software%20Quality.pdf).
+
+*Note that the term "Code Quality" here is scoped to be the readability, understandability, and maintainability of the code. While performance can be considered a legitimate basis for some metrics (e.g. time taken to run certain operations, based on profiling), it is not considered for this purpose.*
+
+## Example of Metrics
+
+### Complexity Metrics
+
+Complexity metrics measures how "complex" methods, classes, packages, etc. are. It is arguably the most useful metric for the largest number of developer because it is the easiest to grasp, the most directly relevant to the coding activity, and applicable to most/all programming languages.
+
+- The most well-known complexity metric is **Cyclomatic Complexity (CC)**, invented by Thomas McCabe in 1976. [Here](http://www.whiteboxtest.com/cyclomatic-complexity.php) is an excellent explanation on how to make use of it to measure your code quality. CC corresponds to the **minimum number of test cases needed to achieve 100% branch coverage**.
+- An alternative, less well-known, harder to calculate complexity metric is **NPath Complexity (NC)**, invented by Brian A. Nejmeh in 1988. NC corresponds to the **minimum number of test cases needed to achieve 100% path coverage**.
+
+> [Here](http://www.literateprogramming.com/mccabe.pdf) is the original paper by McCabe, and [here](http://delivery.acm.org/10.1145/50000/42379/p188-nejmeh.pdf?ip=137.132.250.8&id=42379&acc=ACTIVE%20SERVICE&key=FF6731C4D3E3CFFF%2EBB5EB8D2067C1662%2E4D4702B0C3E38B35%2E4D4702B0C3E38B35&CFID=726592962&CFTOKEN=31567842&__acm__=1486743928_30f8e35ed72667612c6449afe5db3998) is the original paper by Nejmeh.
```

Comment:
As it turns out, the link is generated only for short-term usage (as evident from the "ip" parameter in the web link). Will link to the [citation page](http://dl.acm.org/citation.cfm?doid=42372.42379) instead.